bennu-core/src/main/java/org/fenixedu/bennu/core/domain/User.java

```diff
@@ -167,38 +168,101 @@ public String getEmail() {
+            throw new Error("Please provide proper cryptographic algorithm implementation");
         }
-        byte salt[] = new byte[64];
-        prng.nextBytes(salt);
-        setSalt(BaseEncoding.base64().encode(salt));
-        String hash = Hashing.sha512().hashString(getSalt() + password, Charsets.UTF_8).toString();
-        setPassword(hash);
     }
 
+    /**
+     * Verifies that the given password matches the user's password. Fails with {@link Error} if the necessary cryptographic
+     * algorithm is not present in the environment.
+     * 
+     * @param password the password to verify
+     * @return true if matches, false otherwise
+     */
     public boolean matchesPassword(final String password) {
         if (getPassword() == null) {
-            return true;
+            return false;
+        }
+        if (!getPassword().contains(":")) {
+            final String hash = Hashing.sha512().hashString(getSalt() + password, Charsets.UTF_8).toString();
+            return hash.equals(getPassword());
+        } else {
+            try {
+                String[] params = getPassword().split(":");
+                int iterations = Integer.parseInt(params[1]);
+                byte[] salt = BaseEncoding.base64().decode(params[2]);
+                byte[] hash = BaseEncoding.base64().decode(params[3]);
+                byte[] testHash = pbkdf2(password.toCharArray(), salt, iterations, hash.length);
+                return slowEquals(hash, testHash);
+            } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+                throw new Error("Please provide proper cryptographic algorithm implementation");
+            }
         }
-        final String hash = Hashing.sha512().hashString(getSalt() + password, Charsets.UTF_8).toString();
-        return hash.equals(getPassword());
     }
 
     public Group groupOf() {
         return Group.users(this);
     }
 
+    /**
+     * Compares two byte arrays in length-constant time. This comparison method
+     * is used so that password hashes cannot be extracted from an on-line
+     * system using a timing attack and then attacked off-line.
+     * 
+     * @param a the first byte array
+     * @param b the second byte array
+     * @return true if both byte arrays are the same, false if not
+     */
+    private static boolean slowEquals(byte[] a, byte[] b) {
+        int diff = a.length ^ b.length;
+        for (int i = 0; i < a.length && i < b.length; i++)
+            diff |= a[i] ^ b[i];
+        return diff == 0;
+    }
+
+    /**
+     * Computes the PBKDF2 hash of a password.
+     *
+     * @param password the password to hash.
+     * @param salt the salt
+     * @param iterations the iteration count (slowness factor)
+     * @param bytes the length of the hash to compute in bytes
+     * @return the PBDKF2 hash of the password
+     */
+    private static byte[] pbkdf2(char[] password, byte[] salt, int iterations, int bytes) throws NoSuchAlgorithmException,
```

Comment:
To be more future-proof, and since we are already storing the algorithm in the actual password slot, shouldn't this method also receive the algorithm as an input?